diff --git a/src/Plugin/Condition/EntityBundle.php b/src/Plugin/Condition/EntityBundle.php
index 50d5966..b21af1c 100644
--- a/src/Plugin/Condition/EntityBundle.php
+++ b/src/Plugin/Condition/EntityBundle.php
@@ -106,8 +106,17 @@ class EntityBundle extends ConditionPluginBase implements ConstraintConditionInt
     if (empty($this->configuration['bundles']) && !$this->isNegated()) {
       return TRUE;
     }
-    /** @var \Drupal\Core\Entity\ContentEntityInterface $entity */
-    $entity = $this->getContextValue($this->bundleOf->id());
+
+    $context_value = $this->getContextValue($this->bundleOf->id());
+    if (is_object($context_value) && in_array('Drupal\Core\Entity\ContentEntityInterface', class_implements($context_value))) {
+      $entity = $context_value;
+    }
+    else {
+      $entity = \Drupal::entityTypeManager()
+        ->getStorage($this->bundleOf->id())
+        ->load($context_value);
+    }
+
     return !empty($this->configuration['bundles'][$entity->bundle()]);
   }
 
