diff --git a/core/lib/Drupal/Core/Access/AccessManager.php b/core/lib/Drupal/Core/Access/AccessManager.php
index b7b7c0b..1fa2522 100644
--- a/core/lib/Drupal/Core/Access/AccessManager.php
+++ b/core/lib/Drupal/Core/Access/AccessManager.php
@@ -130,9 +130,9 @@ public function check(RouteMatchInterface $route_match, AccountInterface $accoun
     $result = AccessResult::neutral();
     if (!empty($checks)) {
       $arguments_resolver = $this->argumentsResolverFactory->getArgumentsResolver($route_match, $account, $request);
-      $result = AccessResult::allowed();
+      $result = AccessResult::neutral();
       foreach ($checks as $service_id) {
-        $result = $result->andIf($this->performCheck($service_id, $arguments_resolver));
+        $result = $result->orIf($this->performCheck($service_id, $arguments_resolver));
       }
     }
     return $return_as_object ? $result : $result->isAllowed();
