Index: core/modules/user/src/Plugin/Block/UserLoginBlock.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/modules/user/src/Plugin/Block/UserLoginBlock.php	(date 1489576249000)
+++ core/modules/user/src/Plugin/Block/UserLoginBlock.php	(revision )
@@ -72,7 +72,7 @@
    */
   protected function blockAccess(AccountInterface $account) {
     $route_name = $this->routeMatch->getRouteName();
-    if ($account->isAnonymous() && !in_array($route_name, array('user.login', 'user.logout'))) {
+    if ($account->isAnonymous() && !in_array($route_name, array(/*'user.login', */'user.logout'))) {
       return AccessResult::allowed()
         ->addCacheContexts(['route.name', 'user.roles:anonymous']);
     }
