diff --git a/core/lib/Drupal/Core/Form/ConfirmFormHelper.php b/core/lib/Drupal/Core/Form/ConfirmFormHelper.php
index 5446a3d..4fa567d 100644
--- a/core/lib/Drupal/Core/Form/ConfirmFormHelper.php
+++ b/core/lib/Drupal/Core/Form/ConfirmFormHelper.php
@@ -33,13 +33,21 @@ public static function buildCancelLink(ConfirmFormInterface $form, Request $requ
     $url = NULL;
     // If a destination is specified, that serves as the cancel link.
     if ($query->has('destination')) {
-      $options = UrlHelper::parse($query->get('destination'));
-      // @todo Revisit this in https://www.drupal.org/node/2418219.
-      try {
-        $url = Url::fromUserInput('/' . ltrim($options['path'], '/'), $options);
-      }
-      catch (\InvalidArgumentException $e) {
-        // Suppress the exception and fall back to the form's cancel url.
+      $destination = $query->get('destination');
+      if (!UrlHelper::isExternal($destination)) {
+        $options = UrlHelper::parse($destination);
+        if (strpos($options['path'], '/') === 0) {
+          // WTF now.  it already includes the base URL.
+          // $url = Url::fromUrl('internalwithbase:'. $options['path'], $options);
+        }
+        else {
+          try {
+            // @todo Revisit this in https://www.drupal.org/node/2418219.
+            $url = Url::fromUserInput('/' . ltrim($options['path'], '/'), $options);
+          } catch (\InvalidArgumentException $e) {
+            // Suppress the exception and fall back to the form's cancel url.
+          }
+        }
       }
     }
     // Check for a route-based cancel link.
