diff --git a/src/Entity/Views/GroupContentViewsData.php b/src/Entity/Views/GroupContentViewsData.php
index 62eb94a..0e1d234 100644
--- a/src/Entity/Views/GroupContentViewsData.php
+++ b/src/Entity/Views/GroupContentViewsData.php
@@ -94,7 +94,7 @@ class GroupContentViewsData extends EntityViewsData {
       }
 
       $t_args = [
-        '@entity_type' => $entity_type->getLabel(),
+        '@entity_type' => $entity_type->getLowercaseLabel(),
       ];
 
       // This relationship will allow a group content entity to easily map to a
