diff --git a/docroot/modules/contrib/better_exposed_filters/src/Plugin/views/exposed_form/BetterExposedFilters.php b/docroot/modules/contrib/better_exposed_filters/src/Plugin/views/exposed_form/BetterExposedFilters.php
index 20372c8..638d281 100644
--- a/docroot/modules/contrib/better_exposed_filters/src/Plugin/views/exposed_form/BetterExposedFilters.php
+++ b/docroot/modules/contrib/better_exposed_filters/src/Plugin/views/exposed_form/BetterExposedFilters.php
@@ -354,6 +354,10 @@ public function buildOptionsForm(&$form, FormStateInterface $form_state) {
       if (!empty($filter->options['expose']['label'])) {
         $identifier .= $this->t(' (Filter label: "@fl")', array('@fl' => $filter->options['expose']['label']));
       }
+
+      // Alter the list of available display options for this filter.
+      \Drupal::moduleHandler()->alter('better_exposed_filters_display_options', $display_options, $filter);
+
       $bef_options[$label]['bef_format'] = array(
         '#type' => 'select',
         '#title' => $this->t('Display @identifier exposed filter as', array('@identifier' => $identifier)),
@@ -519,9 +523,6 @@ public function buildOptionsForm(&$form, FormStateInterface $form_state) {
     }
     /* Ends: foreach ($filters as $filter) { */
 
-    // Alter the list of available display options for this filter.
-    \Drupal::moduleHandler()->alter('better_exposed_filters_display_options', $display_options, $filter);
-
     // Add BEF form elements to the exposed form options form.
     $form['bef'] = $bef_options;
   }
