diff --git a/realname.module b/realname.module
index 871804e..68520e3 100644
--- a/realname.module
+++ b/realname.module
@@ -246,7 +246,13 @@ function realname_update(User $account) {
   // Allow modules to react to the realname being updated.
   \Drupal::moduleHandler()->invokeAll('realname_update', array($realname, $account));
 
-  $account->setUsername($realname);
+  $result = \Drupal::database()->select('users_field_data', 'users_field_data')
+    ->fields('users_field_data', array('uid'))
+    ->condition('name', $realname)
+    ->execute()
+    ->fetchCol();
+
+  $account->setUsername(empty($result) ? $realname : $realname . ' - ' . $account->id());
   return $realname;
 }
 
