diff --git a/src/Controller/FormModeManagerController.php b/src/Controller/FormModeManagerController.php
index 6d48bd3..6ee3f31 100644
--- a/src/Controller/FormModeManagerController.php
+++ b/src/Controller/FormModeManagerController.php
@@ -171,7 +171,7 @@ public function addPageTitle($entity_bundle_id, $form_display, EntityTypeInterfa
   /**
    * Checks access for the Form Mode Manager routes.
    *
-   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   * @param string $entity_type
    *   The entity type definition. Useful when a single class is,
    *   used for multiple, possibly dynamic entity types.
    * @param string $form_display
@@ -180,8 +180,8 @@ public function addPageTitle($entity_bundle_id, $form_display, EntityTypeInterfa
    * @return \Drupal\Core\Access\AccessResultInterface
    *   The access result.
    */
-  public function checkAccess(EntityTypeInterface $entity_type, $form_display = 'default') {
-    return AccessResult::allowedIfHasPermission($this->currentUser(), "use {$entity_type->id()}.{$form_display} form mode")
+  public function checkAccess(string $entity_type, $form_display = 'default') {
+    return AccessResult::allowedIfHasPermission($this->currentUser(), "use {$entity_type}.{$form_display} form mode")
       ->cachePerPermissions();
   }
 
diff --git a/src/Routing/FormModeManagerRouteSubscriber.php b/src/Routing/FormModeManagerRouteSubscriber.php
index b1a37b7..eb5d6bd 100644
--- a/src/Routing/FormModeManagerRouteSubscriber.php
+++ b/src/Routing/FormModeManagerRouteSubscriber.php
@@ -157,21 +157,21 @@ protected function getFormModeManagerAddRoute(RouteCollection $collection, Entit
             $collection_route_name = $entity_type_id . '.add_form';
           }
           $route_add = $collection->get($collection_route_name);
-          $route_add->addDefaults(['entity_type' => $entity_type]);
+          $route_add->addDefaults(['entity_type' => $entity_type_id]);
           $route_add->setRequirement('_custom_access', '\Drupal\form_mode_manager\Controller\FormModeManagerController::checkAccess');
           // node.add_page route alter.
           $route_list = $collection->get("$entity_type_id.add_page");
-          $route_list->addDefaults(['entity_type' => $entity_type]);
+          $route_list->addDefaults(['entity_type' => $entity_type_id]);
           $route_list->setRequirement('_custom_access', '\Drupal\form_mode_manager\Controller\FormModeManagerController::checkAccess');
           // entity.node.edit_form route alter.
           $route_edit_form = $collection->get("entity.$entity_type_id.edit_form");
-          $route_edit_form->addDefaults(['entity_type' => $entity_type]);
+          $route_edit_form->addDefaults(['entity_type' => $entity_type_id]);
           $route_edit_form->setRequirement('_custom_access', '\Drupal\form_mode_manager\Controller\FormModeManagerController::checkAccess');
           break;
 
         case 'user':
           $route_add = $collection->get("user.register");
-          $route_add->addDefaults(['entity_type' => $entity_type]);
+          $route_add->addDefaults(['entity_type' => $entity_type_id]);
           $route_add->setRequirement('_custom_access', '\Drupal\form_mode_manager\Controller\FormModeManagerController::checkAccess');
           break;
       }
