diff --git a/src/Plugin/EntityLegal/Redirect.php b/src/Plugin/EntityLegal/Redirect.php
index c9c0071..65234cf 100644
--- a/src/Plugin/EntityLegal/Redirect.php
+++ b/src/Plugin/EntityLegal/Redirect.php
@@ -45,6 +45,11 @@ class Redirect extends EntityLegalPluginBase {
       $event = &$context['event'];
       $response = new TrustedRedirectResponse($entity_url->toString());
       $event->setResponse($response);
+
+      // Remove destination cause the RedirectResponseSubscriber redirects and in some cases it brings redirect loops.
+      $request = $event->getRequest();
+      $request->query->remove('destination');
+      $request->request->remove('destination');
     }
   }
 
